providers/implementations/include/prov/ciphercommon.h

```diff
@@ -88,10 +88,10 @@ struct prov_cipher_ctx_st {
     unsigned int num;
 
     /* The original value of the iv */
-    unsigned char oiv[GENERIC_BLOCK_SIZE];
+    unsigned char oiv[GENERIC_BLOCK_SIZE] __attribute__ ((aligned (GENERIC_BLOCK_SIZE)));
     /* Buffer of partial blocks processed via update calls */
-    unsigned char buf[GENERIC_BLOCK_SIZE];
-    unsigned char iv[GENERIC_BLOCK_SIZE];
+    unsigned char buf[GENERIC_BLOCK_SIZE] __attribute__ ((aligned (GENERIC_BLOCK_SIZE)));
+    unsigned char iv[GENERIC_BLOCK_SIZE] __attribute__ ((aligned (GENERIC_BLOCK_SIZE)));
```

Comment:
There are compilers which we need to support that wouldn't support this syntax. This needs to be handled via a macro that is defined.

Reply:
Please look at how OSSL_UNION_ALIGN is used to achieve the alignment without this compiler specific thing.

Reply:
> Please look at how OSSL_UNION_ALIGN is used to achieve the alignment without this compiler specific thing.

Thanks for this instruction! Before this comment I tried another way learnt from https://github.com/openssl/openssl/blob/221d65ba534d23a240ccadd0c2679b222aae35b1/crypto/ec/curve448/field.h#L25-L33

After this comment I tried to wrap oiv/iv/buf in union like this
```
    union {
        OSSL_UNION_ALIGN;
        unsigned char iv[GENERIC_BLOCK_SIZE];
    } iv; 
```
But when compiling there are errors like `note: expected 'unsigned char *' but argument is of type 'union <anonymous>'`. It seems that using this way would change the API of this struct, and I need to change them all around the code base. Is this acceptable?

Reply:
> But when compiling there are errors like `note: expected 'unsigned char *' but argument is of type 'union <anonymous>'`. It seems that using this way would change the API of this struct, and I need to change them all around the code base. Is this acceptable?

Yes, as this is an internal API, so it can be freely changed. However we might want to discuss whether we prefer this or the way making this alignment available just on gcc/clang.

Reply:
Another option that might be even less invasive and work everywhere would be to move the 1 bit structure members such as the `pad`, `enc`, etc. to the end of the structure.

Reply:
> move the 1 bit structure members such as the `pad`, `enc`, etc. to the end of the structure

I tried so and the result is still unaligned.
```
(gdb) p &((struct prov_cipher_ctx_st*)0x3ff7fca7e0)->oiv
$3 = (unsigned char (*)[16]) 0x3ff7fca84c
```

The problem here is that there are too many unsigned int before these buffers, and unsigned int is 4 bytes.

How about move these three buffers to the beginning of the struct. I think the struct itself should be 8 byte aligned.

Reply:
Yeah, that should work

Reply:
Done so in new commit. The result is satisfying.

```
(gdb) p &((struct prov_cipher_ctx_st*)0x3ff7fca7e0)->oiv
$1 = (unsigned char (*)[16]) 0x3ff7fca7e0
(gdb) p &((struct prov_cipher_ctx_st*)0x3ff7fca7e0)->buf
$2 = (unsigned char (*)[16]) 0x3ff7fca7f0
(gdb) p &((struct prov_cipher_ctx_st*)0x3ff7fca7e0)->iv 
$3 = (unsigned char (*)[16]) 0x3ff7fca800
```